### PR TITLE
FR: Add simple message types validators for the WRP messages

### PR DESCRIPTION
**Pull request by denopink**
## Overview

related to #25, #78, xmidt-org/scytale#88, xmidt-org/talaria#153 and builds on top of #80 and #84 .

### tl;dr
This pr introduces our wrp `simple message types validators` built with our validation framework introduced in #80 and leverages basic spec validators introduced in #84 . Clients can leverage these prebuilt validators to validate their messages.

<details>
<summary> Explanation</summary>

* Validates requirements described at https://xmidt.io/docs/wrp/simple-messages/
* Includes basic spec validation described at xmidt.io/docs/wrp/basics
* Validates all string fields are UTF8

Clients can leverage our prebuilt validators to validate their messages, such as `SimpleEventValidators` and  `SimpleResponseRequestValidators`:
```go
msgv, err := NewTypeValidator(
	// Validates found msg types
	map[MessageType]Validator{
		SimpleEventMessageType:           SimpleEventValidators(),
		SimpleRequestResponseMessageType: SimpleResponseRequestValidators(),
	},
	// Validates unfound msg types
	AlwaysInvalid())
if err != nil {
	return
}

var (
	expectedStatus                  int64 = 3471
	expectedRequestDeliveryResponse int64 = 34
	expectedIncludeSpans            bool  = true
)
foundErrFailure := msgv.Validate(Message{
	Type: SimpleRequestResponseMessageType,
	// Missing scheme
	Source: "external.com",
	// Invalid Mac
	Destination:             "MAC:+++BB-44-55",
	TransactionUUID:         "DEADBEEF",
	ContentType:             "ContentType",
	Accept:                  "Accept",
	Status:                  &expectedStatus,
	RequestDeliveryResponse: &expectedRequestDeliveryResponse,
	Headers:                 []string{"Header1", "Header2"},
	Metadata:                map[string]string{"name": "value"},
	Spans: [][]string{
		// // Invalid length
		{},
		// Invalid length
		{"3"},
		// Invalid 'start time', 'duration' and 'status' components
		{"parent", "name", "not start time", "not duration", "not status"},
		// Invalid 'parent' and 'name' components
		{"1234", "1234", "1234", "1234", "1234"},
	},
	IncludeSpans: &expectedIncludeSpans,
	Path:         "/some/where/over/the/rainbow",
	Payload:      []byte{1, 2, 3, 4, 0xff, 0xce},
	ServiceName:  "ServiceName",
	// Not UFT8 URL string
	URL:        "someURL\xed\xbf\xbf.com",
	PartnerIDs: []string{"foo"},
	SessionID:  "sessionID123",
}) // Found error
foundErrSuccess1 := msgv.Validate(Message{
	Type:        SimpleRequestResponseMessageType,
	Source:      "MAC:11:22:33:44:55:66",
	Destination: "MAC:11:22:33:44:55:61",
}) // Found success
foundErrSuccess2 := msgv.Validate(Message{
	Type:   SimpleEventMessageType,
	Source: "MAC:11:22:33:44:55:66",
	// Invalid Destination
	Destination: "invalid:a-BB-44-55",
}) // Found error
unfoundErrFailure := msgv.Validate(Message{Type: CreateMessageType}) // Unfound error
fmt.Println(foundErrFailure == nil, foundErrSuccess1 == nil, foundErrSuccess2 == nil, unfoundErrFailure == nil)
// Output: false true false false
```

</details>

<details>
<summary>Type of Change(s)</summary>

- Non-breaking Enhancement
- All new and existing tests passed.

</details>

### CODE REVIEW

File: simpleMessageTypes_validator.go [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"go.uber.org/multierr"
+)
+
+var (
+	ErrorNotSimpleResponseRequestType = errors.New("not simple response request message type")
```
Comment (johnabass) [inline simpleMessageTypes_validator.go]: We should have a concrete error type for these.  It should identify which field, along with a message that can be set by the validator.  This will help clients take corrective action or to selectively ignore validation errors when able to do so.

Something like this can work:

```
type ValidatorError struct {
    // Fields are the relevant fields involved in this error.  This can be empty.
    Fields []reflect.StructField

    // Err is the cause of the error, e.g. a parsing problem.  This can be nil, but
    // either Err or Message must be set.
    Err error

    // Message is a validation message.  This can be empty, but either Err
    // or Message must be set.  Both Err and Message can be set if the validator
    // wants to communicate something beyond the Err cause.
    Message string
}

// so errors.As and errors.Is can work
func (ve *ValidatorError) Unwrap() error {
    return ve.Err
}

func (ve *ValidatorError) Error() string {
    var o strings.Builder
    o.WriteString("Validator error")

    if len(ve.Fields) > 0 {
        o.WriteString(" [")
        for i, f := range ve.Fields {
            if i > 0 {
                o.WriteRune(',')
            }

            o.WriteString(f.Name)
        }

        o.WriteRune(']')
    }

    if ve.Err != nil {
        o.WriteString(" err=")
        o.WriteString(ve.Err.String()) 
    } else if len(ve.Message) > 0 {
        o.WriteString(" msg=")
        o.WriteString(ve.Message) 
    }

    return o.String()
}
```

File: simpleMessageTypes_validator.go [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+// SimpleResponseRequestValidators returns a WRP validator that ensures messages are valid based on
+// each validator in the list. SimpleResponseRequestValidators validates the following:
+// UTF8 (all string fields), MessageType is valid, Source, Destination, Spans, MessageType is of
+// SimpleRequestResponseMessageType.
+func SimpleResponseRequestValidators() Validators {
+	return Validators{SpecValidators(), SimpleResponseRequestTypeValidator(), SpansValidator()}
+}
+
+// SimpleResponseRequestTypeValidator returns a WRP validator that takes messages and validates their Type is of SimpleRequestResponseMessageType.
+func SimpleResponseRequestTypeValidator() ValidatorFunc {
```
Comment (johnabass) [inline simpleMessageTypes_validator.go]: This doesn't have to be a function returning the validator.  It can just be the validator.

e.g. `SimpleResponseRequestTypeValidator(m Message) error { ... }`

See my comment below about interfaces.  Under that approach, a code would use this validator by doing `ValidatorFunc(SimpleResponseRequestTypeValidator)`.

File: simpleMessageTypes_validator.go [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+var (
+	ErrorNotSimpleResponseRequestType = errors.New("not simple response request message type")
+	ErrorNotSimpleEventType           = errors.New("not simple event message type")
+	ErrorInvalidSpanLength            = errors.New("invalid span length")
+	ErrorInvalidSpanFormat            = errors.New("invalid span format")
+)
+
+// spanFormat is a simple map of allowed span format.
+var spanFormat = map[int]string{
+	0: "parent",
```
Comment (mtrinh11) [inline simpleMessageTypes_validator.go:38]: do we want a comment about the parent here too?

File: simpleMessageTypes_validator.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+		// Spans consist of individual Span(s), arrays of timing values.
+		for _, s := range m.Spans {
+			if len(s) != len(spanFormat) {
+				err = multierr.Append(err, ErrorInvalidSpanLength)
+				continue
+			}
+
+			for i, j := range spanFormat {
+				switch j {
+				// Any nonemtpy string is valid
```
Comment (mtrinh11) [inline simpleMessageTypes_validator.go]: spelling of "nonempty"

**Review by mtrinh11 — APPROVED** [review]
Super duper minor comments, otherwise lgtm!

File: simpleMessageTypes_validator.go [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+				}
+			}
+		}
+
+		return err
+	}
+}
+
+// SimpleEventTypeValidator returns a WRP validator that takes messages and validates their Type is of SimpleEventMessageType.
+func SimpleEventTypeValidator() ValidatorFunc {
```
Comment (johnabass) [inline simpleMessageTypes_validator.go]: Validators should really implement a common interface.  The usual pattern of providing a function type that implements that interface works well.

```
type Validator interface {
    Validate(m *Message) error
}

type ValidatorFunc func(*Message) error

func (vf ValidatorFunc) Validate(m *Message) error { return vf(m) }

// For data structures that use validators, providing (2) methods for each of
// the above types makes things easier on client code
type Validators []Validator
func (vs Validators) Add(v ...Validator) { ... }
func (vs Validators) AddFunc(f ...ValidatorFunc) { ... } // now a client doesn't have to cast
```

That approach allows for custom validators to more easily have state, since a client can provide a `struct` that has that state.  Trivial validators, like the ones in this package, can just be functions.

**Review by johnabass — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 6]
```diff
@@ -0,0 +1,24 @@
+# SPDX-FileCopyrightText: 2022 Comcast Cable Communications Management, LLC
```
Comment (denopink) [inline .github/workflows/ci.yml]: `git rebase upstream/main`, these are changes from #88

File: .github/workflows/push.yml [hunk 7]
```diff
@@ -1,107 +0,0 @@
-name: CI
```
Comment (denopink) [inline .github/workflows/push.yml]: `git rebase upstream/main`, these are changes from #88

File: .github/workflows/release.yml [hunk 8]
```diff
@@ -1,34 +0,0 @@
-name: release
```
Comment (denopink) [inline .github/workflows/release.yml]: `git rebase upstream/main`, these are changes from #88

File: .github/workflows/tag.yml [hunk 9]
```diff
@@ -1,28 +0,0 @@
-name: tag
```
Comment (denopink) [inline .github/workflows/tag.yml]: `git rebase upstream/main`, these are changes from #88

File: .github/workflows/updater.yml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+# SPDX-FileCopyrightText: 2022 Comcast Cable Communications Management, LLC
```
Comment (denopink) [inline .github/workflows/updater.yml]: `git rebase upstream/main`, these are changes from #88

File: .sonar-project.properties [hunk 11]
```diff
@@ -1,34 +0,0 @@
-# Reference:
```
Comment (denopink) [inline .sonar-project.properties]: `git rebase upstream/main`, these are changes from #88

File: README.md [hunk 12]
```diff
@@ -3,13 +3,14 @@
 wrp-go provides a library implementing the [Web Routing Protocol](https://github.com/xmidt-org/wrp-c/wiki/Web-Routing-Protocol)
```
Comment (denopink) [inline README.md]: `git rebase upstream/main`, these are changes from #88

File: validator.go [hunk 13]
```diff
@@ -19,16 +19,98 @@ package wrp
...
+	// Either Err or Message must be set.
+	Err error
+
+	// Message is a validation message in case the validator wants
+	// to communicate something beyond the Err cause.
+	Message string
+}
+
+// Is reports whether any error in e.err's chain matches target.
+func (ve ValidatorError) Is(target error) bool {
```
Comment (johnabass) [inline validator.go]: This method isn't necessary.  Because you've supplied `Unwrap`, the plain old `errors.Is(...)` will work.

File: validator.go [hunk 14]
```diff
@@ -19,16 +19,98 @@ package wrp
...
+		o.WriteString(" msg=")
+		o.WriteString(ve.Message)
+	}
+
+	return o.String()
+}
+
+func NewValidatorError(err error, sf []string, m string) ValidatorError {
+	// Either Err or Message must be set.
+	if err == nil && len(m) == 0 {
```
Comment (johnabass) [inline validator.go]: Returning an error from creating an error is a bit redundant.  It's ok if an error object's method panic due to incorrect setup.

File: validator.go [hunk 15]
```diff
@@ -19,16 +19,98 @@ package wrp
...
+
+	if len(ve.Message) > 0 {
+		o.WriteString(" msg=")
+		o.WriteString(ve.Message)
+	}
+
+	return o.String()
+}
+
+func NewValidatorError(err error, sf []string, m string) ValidatorError {
```
Comment (johnabass) [inline validator.go]: For structs that have all exported members, it's not strictly necessary to have a constructor function like this.  Code like this is very simple for clients:

```
return ValidatorError{
    Message: "oh no!",
    Err: cause,
}
```

File: validator.go [hunk 16]
```diff
@@ -19,16 +19,98 @@ package wrp
...
+	// Either Err or Message must be set.
+	if err == nil && len(m) == 0 {
+		return ValidatorError{Err: errorInvalidValidatorError}
+	}
+
+	var badFields string
+	verr := ValidatorError{Err: err, Fields: make([]reflect.StructField, 0, len(sf)), Message: m}
+	// Fields must exist.
+	for _, f := range sf {
+		ft, ok := messageReflectType.FieldByName(f)
```
Comment (johnabass) [inline validator.go]: I realized that my suggestion of using `reflect.StructField` won't work well in the long-term.  That's because we may need to refer to fields in the `payload`, which might be JSON or some other arbitrary format.

We should probably just simplify this and may the fields a `'[]string`.  Sorry I led you astray!

**Review by johnabass — CHANGES_REQUESTED** [review]

**Review by johnabass — APPROVED** [review]
